transaction/interfaces.py

```diff
@@ -356,6 +356,74 @@ def getAfterCommitHooks():
         by a top-level transaction commit.
         """
 
+    def addBeforeAbortHook(hook, args=(), kws=None):
+        """Register a hook to call before the transaction is abortted.
+
+        The specified hook function will be called after the transaction's
+        abort method has been called, but before the abort process has been
+        started.  The hook will be passed the specified positional (`args`)
+        and keyword (`kws`) arguments.  `args` is a sequence of positional
+        arguments to be passed, defaulting to an empty tuple (no positional
+        arguments are passed).  `kws` is a dictionary of keyword argument
+        names and values to be passed, or the default None (no keyword
+        arguments are passed).
+
+        Multiple hooks can be registered and will be called in the order they
+        were registered (first registered, first called).  This method can
+        also be called from a hook:  an executing hook can register more
+        hooks.  Applications should take care to avoid creating infinite loops
+        by recursively registering hooks.
+
+        Hooks are called only for a top-level abort. If the
```

Comment:
I wonder if those are the best semantics. If someone wants to catch all possible aborts, they need to register two hooks: one for top-level aborts, and one for after commit, checking for `False`. It seems like it would be easier to also call the abort hooks from `_cleanup` (error handling during commit) so that all cases where a transaction is aborted can be handled with a single hook. Of course, that might raise the possibility of wanting to know whether the abort was explicit or implicit...

I'm not sure I have a strong opinion about this, but I figured it was worth discussing.

Reply:
@jamadden 
> I wonder if those are the best semantics. If someone wants to catch all possible aborts, they need to register two hooks: one for top-level aborts, and one for after commit, checking for `False`....

I am not yet sure - but maybe, the `abort` hook would be sufficient: when I remember right, then a transaction with commit problems enters a "DOOMED" state and must be explicitly "aborted". Should this be true, then we could expect, that typically a top level `abort` follows a commit with problems.

Reply:
I have verified: in case of a commit problem, the transaction enters the `COMMITFAILED` state and becomes virtually unusable. The natural thing to do in this case is to `abort` the transaction. in order to get a new (clean) one. I have changed the documentation accordingly.